### PR TITLE
Add dynamic return type for `get_permalink()`/`get_the_permalink()`

### PR DESCRIPTION
**Pull request by IanDelMar**
Adds dynamic function return type extension for `get_permalink()` and `get_the_permalink()`.

### CODE REVIEW

File: src/GetPermalinkDynamicFunctionReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            $type = $scope->getType($args[0]->value);
+
+            // Called with a WP_Post instance
+            if ($type instanceof ObjectType && $type->isInstanceOf('WP_Post')->yes()) {
+                return new StringType();
+            }
+        }
+
+        // When called without arguments or with a $type that isn't a WP_Post instance, return default return type
+        return ParametersAcceptorSelector::selectFromArgs(
```
Comment (herndlm) [inline src/GetPermalinkDynamicFunctionReturnTypeExtension.php:51]: It should be also possible to just return `null` instead if you want to simplify this

**Review by herndlm — COMMENTED** [review]
Also small idea for an improvement - what do you think of letting it return false in case the arg is not an object AND not an integer? That could help find some false usages of those functions :)

File: src/GetPermalinkDynamicFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @see https://developer.wordpress.org/reference/functions/get_permalink/
+     * @see https://developer.wordpress.org/reference/functions/get_the_permalink/
+     */
+    public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+    {
+        $args = $functionCall->getArgs();
+
+        if (count($args) > 0) {
```
Comment (szepeviktor) [inline src/GetPermalinkDynamicFunctionReturnTypeExtension.php:41]:
```suggestion
        if (count($args) !== 0) {
```

**Comment by szepeviktor** [comment]
@IanDelMar Thank you for working on a Saturday! 📅

**Comment by IanDelMar** [comment]
> Also small idea for an improvement - what do you think of letting it return false in case the arg is not an object AND not an integer? That could help find some false usages of those functions :)

PHPStan knows that the first argument has to be of type `int` or `WP_Post` because of the stubs file.

**Comment by IanDelMar** [comment]
> Also small idea for an improvement - what do you think of letting it return false in case the arg is not an object AND not an integer? That could help find some false usages of those functions :)

PHPStan knows that the first argument must be of type `int` or `WP_Post` because of the stubs file. So there is an error if you use a non integer, non WP_Post argument. Using a string type argument gives `Parameter #1 $post of function get_permalink expects int|WP_Post, string given.`

Using the return value as an argument for a function that expects a string, e.g. `str_replace( 'a string', 'a new string', get_permalink( 'not expected string' ) )` gives `Parameter #3 $subject of function str_replace expects array|string, string|false given.`

But maybe it is helpful for some users to have `Parameter #3 $subject of function str_replace expects array|string, false given.` instead.

File: src/GetPermalinkDynamicFunctionReturnTypeExtension.php (hunk 1 above)
Comment (szepeviktor) [reply src/GetPermalinkDynamicFunctionReturnTypeExtension.php:51]: @IanDelMar Could you try using `null` here?

Comment (IanDelMar) [reply src/GetPermalinkDynamicFunctionReturnTypeExtension.php:51]: Can you give an example when `get_permalink()` returns `null`?

Comment (IanDelMar) [reply src/GetPermalinkDynamicFunctionReturnTypeExtension.php:51]: Ah, now I see. You mean returning `null` is equivalent to returning the default type(s)?

Comment (IanDelMar) [reply src/GetPermalinkDynamicFunctionReturnTypeExtension.php:51]: > It should be also possible to just return `null` instead if you want to simplify this

True. Unit tests pass.

Comment (szepeviktor) [reply src/GetPermalinkDynamicFunctionReturnTypeExtension.php]: Thank you.
Please do not care about `SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter`

**Comment by herndlm** [comment]
Oh, of course, makes sense. Thx!

**Pull request merged**